### PR TITLE
[engine][checkpoint] checkpoint base classes

### PR DESCRIPTION
**Pull request by ashulin**
## Purpose of this pull request
add checkpoint base classes

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        CheckpointCoordinatorConfiguration that = (CheckpointCoordinatorConfiguration) o;
+        return checkpointInterval == that.checkpointInterval
+            && checkpointTimeout == that.checkpointTimeout
+            && maxConcurrentCheckpoints == that.maxConcurrentCheckpoints
+            && tolerableFailureCheckpoints == that.tolerableFailureCheckpoints;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorConfiguration.java:80]: Why not use lombok annotations to generate code?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CompletedCheckpoint.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        this.taskStates = taskStates;
+    }
+
+    @Override
+    public long getCheckpointId() {
+        return this.checkpointId;
+    }
+
+    @Override
+    public long getPipelineId() {
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CompletedCheckpoint.java:59]: as above

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
+        out.writeObject(taskInfo);
+        out.writeByteArray(states);
+    }
+
+    @Override
+    protected void readInternal(ObjectDataInput in) throws IOException {
+        taskInfo = in.readObject(CheckpointBarrier.class);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java]: Use TaskInfo.class?  
TaskInfo need to implement the Serializable interface?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/TaskCompletedOperation.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
+        out.writeObject(taskInfo);
+    }
+
+    @Override
+    protected void readInternal(ObjectDataInput in) throws IOException {
+        taskInfo = in.readObject(CheckpointBarrier.class);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/TaskCompletedOperation.java]: Use TaskInfo.class ?

Comment (ashulin) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/TaskCompletedOperation.java]: done

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java (hunk 3 above)
Comment (ashulin) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java]: done

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorConfiguration.java (hunk 1 above)
Comment (ashulin) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorConfiguration.java:80]: I don't recommend or oppose lombok in open source components;
At present, seatunnel also uses lombok for some people, while others do not;
Maybe we need to discuss this?

Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorConfiguration.java:80]: > I don't recommend or oppose lombok in open source components; 

agree with you.

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public CheckpointAckOperation(TaskInfo taskInfo, byte[] states) {
+        this.taskInfo = taskInfo;
+        this.states = states;
+    }
+
+    @Override
+    public int getClassId() {
+        return OperationDataSerializerHook.SUBMIT_OPERATOR;
```
Comment (EricJoy2048) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java]: `OperationDataSerializerHook.SUBMIT_OPERATOR` is used for `SubmitJobOperation`

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointFinishedOperation.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                                       long pipelineId,
+                                       long checkpointId) {
+        this.jobId = jobId;
+        this.pipelineId = pipelineId;
+        this.checkpointId = checkpointId;
+    }
+
+    @Override
+    public int getClassId() {
+        return OperationDataSerializerHook.SUBMIT_OPERATOR;
```
Comment (EricJoy2048) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointFinishedOperation.java]: same as above

**Comment by EricJoy2048** [comment]
Operator extend `AsyncOperator` only the Server method return a `CompleteFuture` , otherwise we only need extend `Operator` class like `PrintMessageOperator`

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java (hunk 5 above)
Comment (ashulin) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointAckOperation.java]: done

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointFinishedOperation.java (hunk 6 above)
Comment (ashulin) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/operation/CheckpointFinishedOperation.java]: done

**Comment by ashulin** [comment]
> Operator extend `AsyncOperator` only the Server method return a `CompleteFuture` , otherwise we only need extend `Operator` class like `PrintMessageOperator`

According to the current design, the steps that need to be synchronized for checkpoint are inside the task.

**Comment by EricJoy2048** [comment]
> > Operator extend `AsyncOperator` only the Server method return a `CompleteFuture` , otherwise we only need extend `Operator` class like `PrintMessageOperator`
> 
> According to the current design, the steps that need to be synchronized for checkpoint are inside the task.

Yes, If you need a synchronized call, you can reference the `PrintMessageOperator`.

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**